### PR TITLE
Enforce a maximum of 6 day intervals for RetryOptions

### PR DESCRIPTION
**Pull request by ConnorMcMahon**
Our current version of the storage SDK limits our timer functionality to a maximum of 7 days. We already had these checks for timers, but the same limitations apply to our retry intervals. Without enforcing these limits, customers could get permenantly stuck, and it was difficult to debug why. This silently sets the interval to our maximum when the customer attempts to use intervals larger than our maximum.

Addresses issue #944.

### CODE REVIEW

**Comment by ConnorMcMahon** [comment]
This unfortunately could break in flight orchestrations that may or may not otherwise execute in a normal fashion. This is because we now throw an exception as soon as we encounter a RetryOptions that has a greater `MaxRetryInterval` than 6 days, where in all prior releases we default that value to `TimeSpan.MaxValue`.

Because of this, it may be better to just log a warning and just edit the value to 6 days instead of throwing an exception message. It's less clear, but it won't break anyone.

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs [hunk 1]
```diff
@@ -422,6 +422,18 @@ public override void ContinueAsNew(object input, bool preserveUnprocessedEvents)
         {
             this.ThrowIfInvalidAccess();
 
+            // These check can be removed once the storage provider supports extended timers.
```
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs]: Debatably I can remove this comment both here and in the Timers section, as it is highly unlikely that we will ever update the Storage SDK in Durable 1.0.

File: src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs [hunk 2]
```diff
@@ -422,6 +422,18 @@ public override void ContinueAsNew(object input, bool preserveUnprocessedEvents)
...
+            // https://github.com/Azure/azure-functions-durable-extension/issues/14
+            if (retryOptions.FirstRetryInterval > TimeSpan.FromDays(MaxTimerDurationInDays))
+            {
+                throw new ArgumentException($"Retry intervals must not exceed {MaxTimerDurationInDays} days.", nameof(retryOptions.FirstRetryInterval));
+            }
+
+            if (retryOptions.MaxRetryInterval > TimeSpan.FromDays(MaxTimerDurationInDays))
+            {
+                throw new ArgumentException($"Retry intervals must not exceed {MaxTimerDurationInDays} days.", nameof(retryOptions.MaxRetryInterval));
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs]: As you mentioned, this is a breaking change, which we can't allow for Durable 1.x. I think the best path forward is to  just set (overwrite) large retry intervals with our 6 day maximum. In the PR for Durable 2.x, we can throw here and/or throw in the property setters.

File: src/WebJobs.Extensions.DurableTask/RetryOptions.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@ namespace Microsoft.Azure.WebJobs
     public class RetryOptions
     {
         private readonly DurableTaskCore.RetryOptions retryOptions;
+        private readonly TimeSpan maxStorageTimeSpan = TimeSpan.FromDays(DurableOrchestrationContext.MaxTimerDurationInDays);
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/RetryOptions.cs]: Let's make this `static` since it's the same for all instances.

File: test/FunctionsV1/WebJobs.Extensions.DurableTask.Tests.V1.csproj [hunk 4]
```diff
@@ -11,7 +11,7 @@
   <ItemGroup>
     <PackageReference Include="FluentAssertions" Version="4.19.4" />
     <PackageReference Include="Microsoft.Azure.DurableTask.AzureStorage" Version="1.6.3" />
-    <PackageReference Include="Microsoft.Azure.WebJobs" Version="2.3.0" />
+    <PackageReference Include="Microsoft.Azure.WebJobs" Version="2.2.0" />
```
Comment (cgillum) [inline test/FunctionsV1/WebJobs.Extensions.DurableTask.Tests.V1.csproj]: Was this change intentional?

File: src/WebJobs.Extensions.DurableTask/RetryOptions.cs [hunk 5]
```diff
@@ -24,6 +25,7 @@ public class RetryOptions
         public RetryOptions(TimeSpan firstRetryInterval, int maxNumberOfAttempts)
         {
             this.retryOptions = new DurableTaskCore.RetryOptions(firstRetryInterval, maxNumberOfAttempts);
+            this.MaxRetryInterval = TimeSpan.FromDays(DurableOrchestrationContext.MaxTimerDurationInDays);
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/RetryOptions.cs]: Looks like we can reuse your field declared on line 15 here.

File: test/FunctionsV1/WebJobs.Extensions.DurableTask.Tests.V1.csproj (hunk 4 above)
Comment (ConnorMcMahon) [reply test/FunctionsV1/WebJobs.Extensions.DurableTask.Tests.V1.csproj]: Yes, there was a warning in the test project because the src code uses 2.2.0.

Comment (ConnorMcMahon) [reply test/FunctionsV1/WebJobs.Extensions.DurableTask.Tests.V1.csproj]: Could go in a separate PR if we want to keep this commit purely what is described in it.

Comment (cgillum) [reply test/FunctionsV1/WebJobs.Extensions.DurableTask.Tests.V1.csproj]: Oh - I didn't realize this was the test project. I have no problem with taking this change in the current PR.

**Comment by ConnorMcMahon** [comment]
Need to address this for V2 before release, especially since we would ideally throw an exception if the customer does it (breaking change). Considering this is an Azure Storage specific limitation, was thinking of handling it in PR #974, or in a PR after that is merged.

**Comment by cgillum** [comment]
Oh right - this is storage provider-specific. I guess it's not a simple port then.
